UVC Video encoder (zephyr,videoenc) + H264 support + STM32 STM32N6 conf files

avolmat-st opened the pull request:
Allow creating a pipeline as follow
       camera receiver -> encoder -> uvc

I post this PR while there are still some points to improve (cf hardcoded stuff detailed below) in order to get some first feedbacks. Since this depends on the UVC PR (PR #93192) and DCMIPP UVC PR (PR #94562), there are lots of commits in this PR. However **only the LAST 8 COMMITS is relevant for this PR.**

If the chosen zephyr,videoenc is available, the sample will pipe the camera receiver to the encoder and then the UVC device instead of directly the camera receiver to the UVC.

In order to making the change as simple as possible, the source device of the UVC device is renamed uvc_src_dev instead of video_dev previously since, depending on the configuration, the UVC source might be either the video_dev or the encoder_dev.

Current implementation has several points hardcoded for the time being:
    1. intermediate pixel format between the camera receiver and encoder
       is set to NV12. This is temporary until proper analysis of video_dev
       caps and encoder caps is done, allowing to select the common
       format of the two devices.
    2. it is considered that encoder device do NOT perform any resolution
       change and that encoder output resolution is directly based on the
       camera receiver resolution.  Thanks to this, UVC exposed formats
       are thus the encoder output pixel format & camera receiver
       resolutions.
    
This has been tested using the STM32N6-DK and the JPEG codec, leading to the following pipe:
    IMX335 -> CSI/DCMIPP -> JPEG -> UVC

compiled via:

```
west build -p -b stm32n6570_dk//sb --shield st_b_cams_imx_mb1854 samples/subsys/usb/uvc/ -DCONFIG_USE_STM32_MW_ISP=y -DFILE_SUFFIX=jpeg
```
And also VENC codec, leading to the following pipe:
     IMX335 -> CSI/DCMIPP -> VENC -> UVC
     
compiled via:

```
west build -p -b stm32n6570_dk//sb --shield st_b_cams_imx_mb1854 samples/subsys/usb/uvc/ -DCONFIG_USE_STM32_MW_ISP=y -DFILE_SUFFIX=venc
```
     
     and can be displayed on linux machine via gst-launch command:

```
gst-launch-1.0 v4l2src device=/dev/videoXX ! 'video/x-h264,width=1920,height=1080' ! h264parse ! avdec_h264 ! videoconvert ! autovideosink
```

Commit:
samples: usb: uvc: add video encoder support

Allow creating a pipeline as follow
   camera receiver -> encoder -> uvc

If the chosen zephyr,videoenc is available, the sample will pipe
the camera receiver to the encoder and then the UVC device instead
of directly the camera receiver to the UVC.

Current implementation has several points hardcoded for the time
being:
1. intermediate pixel format between the camera receiver and encoder
   is set to NV12. This shouldn't be hardcoded and should instead be
   discovered as a commonly capable format from the encoder / video dev
2. it is considered that encoder device do NOT perform any resolution
   change and that encoder output resolution is directly based on the
   camera receiver resolution.  Thanks to this, UVC exposed formats
   are thus the encoder output pixel format & camera receiver
   resolutions.

Signed-off-by: Alain Volmat <alain.volmat@foss.st.com>

Commit:
usb: usbd_uvc: add frame_based support (currently only H264)

The frame_based descriptors differ from the frame descriptors
in that there is no dwMaxVideoFrameBufferSize field.
In order to do that, add a new uvc_frame_based_discrete_descriptor
structure to be used to fill in proper information into the
frame descriptor. In addition to that, a new format descriptor
is also added for frame based transfer.

Signed-off-by: Alain Volmat <alain.volmat@foss.st.com>

josuah reviewed (dismissed):
Thanks for this submission. It makes sense for me as it is, and AFAI the FIXME are for planning future infrastructure of the video area, not for this particular video sample

Commit:
samples: video: tcpserversink: check video_enqueue/dequeue return values

Add proper check of the return value of video_enqueue / video_dequeue.

Signed-off-by: Alain Volmat <alain.volmat@foss.st.com>

jfischer-no commented on samples/subsys/usb/uvc/src/main.c:
```diff
@@ -73,20 +73,62 @@ static void app_add_format(uint32_t pixfmt, uint16_t width, uint16_t height, boo
+	{ .width = 854,		.height = 480,	},	/* WVGA */
+	{ .width = 640,		.height = 480,	},	/* VGA */
+	{ .width = 320,		.height = 240,	},	/* QVGA */
+	{ .width = 160,		.height = 120,	},	/* QQVGA */
```
The default is 5, how many "constrained" devices that Zephyr aims to support can handle UHD, FHD, SXGA, HD, and SVGA and on what frame rate? Should it be sorted ascending?

jfischer-no commented on samples/subsys/usb/uvc/boards/stm32n6570_dk.conf:
```diff
@@ -0,0 +1,3 @@
+# Enough for 2592x1944 YUYV frames
+CONFIG_VIDEO_BUFFER_POOL_SZ_MAX=10100000
+CONFIG_MAIN_STACK_SIZE=2048
```
These are three identical configuration overlays. And I fear that would be just the beginning of the madness. I would accept more sophisticated configuration overlay for one or other board, as an example, but I am absolutely against trivial overlay files. You have to find another way to configure your boards. You can pass these configuration to the build command and add documentation to the sample how to do it.

jfischer-no commented on samples/subsys/usb/uvc/src/main.c:
```diff
@@ -137,6 +155,25 @@ int main(void)
 	struct usbd_context *sample_usbd;
 	struct video_buffer *vbuf;
 	struct video_format fmt = {0};
+#if DT_HAS_CHOSEN(zephyr_videoenc)
```
This should be kept simple and maintainable. `#if DT_HAS_CHOSEN(zephyr_videoenc)` contradicts that, you need to rework these changes.

avolmat-st replied:
reworked, only few DT_HAS_CHOSEN(zephyr_videoenc) at the head of the sample.

jfischer-no commented on samples/subsys/usb/uvc/src/main.c:
```diff
@@ -260,9 +381,27 @@ int main(void)
+				return ret;
+			}
+		}
+#endif
```
Please move them to specific function.

jfischer-no commented on samples/subsys/usb/uvc/boards/stm32n6570_dk_jpeg.overlay:
```diff
@@ -0,0 +1,15 @@
+	uvc: uvc {
+		compatible = "zephyr,uvc-device";
+	};
+};
```
Same as for other board specific overlay files, I do not see anything platform specific here, so it could be samples/subsys/usb/uvc/video_enc_jpeg.overlay but not a board DTS overlay.

josuah replied:
I think `jpeg` is the name of an STM32 JPEG encoder/decoder.

On devices that have a JPEG encoder only, this would be named `jpegenc` (or something more specific).
On devices that have a JPEG decoder only, this would be named `jpegdec` (or something more specific).

But *if* this happen then this could be modified at this time (maybe it's always `jpeg` that can do both encoding+decoding).

jfischer-no commented on samples/subsys/usb/uvc/boards/stm32n6570_dk_stm32n657xx_fsbl_venc.overlay:
```diff
@@ -0,0 +1,15 @@
+	uvc: uvc {
+		compatible = "zephyr,uvc-device";
+	};
+};
```
Same as for other board specific overlay files, I do not see anything platform specific here, so it could be samples/subsys/usb/uvc/video_enc_venc.overlay but not a board DTS overlay.

josuah replied:
ST chose to name their video encoder VENC, but most hardware vendor also do, so that might as well be a suitable default name.

avolmat-st replied:
What would you think of following approach (based on https://github.com/zephyrproject-rtos/zephyr/pull/97425#discussion_r2449132406)

1. addition of **zephyr_h264enc** and **zephyr_jpegenc** which allows for supported platforms to point out their h264 video encoder or jpeg video encoder. This is done in an approach similar to already existing **zephyr_i2c** or **zephyr_udc**. The SW / HW in action when dealing with H264 or JPEG encoder can be different, and for some platform, there is actually both of them (STM32N6).

2. introduce suffixed sample overlay / conf files, for JPEG & H264. So, prj_jpegenc.conf / app_jpegenc.overlay and similar thing for h264enc as well. Compare to prj.conf, each suffixed conf file is also adding the CONFIG_VIDEO_ENCODER_JPEG=y (or CONFIG_VIDEO_ENCODER_H264=y) and overlay is pointing the **zephyr,videoenc** chosen to the video encoder that we want to use, either zephy_jpegenc or zephyr_h264enc.

This can then be selected via -DFILE_SUFFIX="jpegenc" or -DFILE_SUFFIX="h264enc" at build of the uvc sample.

Commit:
video: addition of video_transfer_buffer helper function

Addition of a helper function which takes care of dequeue
from a source video device and queue into a sink video device.

Signed-off-by: Alain Volmat <alain.volmat@foss.st.com>

Commit:
samples: usb: uvc: use video_transfer_buffer helper function

Replace video_dequeue / video_enqueue buffer exchange code by
video_transfer_buffer helper function.

Signed-off-by: Alain Volmat <alain.volmat@foss.st.com>

josuah commented on samples/subsys/usb/uvc/src/main.c:
```diff
@@ -23,6 +23,19 @@ const static struct device *const video_dev = DEVICE_DT_GET(DT_CHOSEN(zephyr_cam
+static struct video_caps videoenc_in_caps = {.type = VIDEO_BUF_TYPE_INPUT};
+static struct video_caps videoenc_out_caps = {.type = VIDEO_BUF_TYPE_OUTPUT};
+const struct device *uvc_src_dev = videoenc_dev;
+struct video_caps *uvc_src_caps = &videoenc_out_caps;
```
For this as well as some other comments, this is provided through this PR to a different video sample, to further be propagated to other samples that use video configuration:
- https://github.com/zephyrproject-rtos/zephyr/pull/96072

avolmat-st replied:
@josuah for the time being I only modified the sample based on the comment received. Added a function to get the uvc source device. (and caps), I didn't modify the way the sample originally do, aka having caps in global variable. That's one point of improvement indeed.

josuah commented on samples/drivers/video/tcpserversink/src/main.c:
```diff
@@ -131,7 +131,10 @@ int configure_encoder(void)
+	if (video_enqueue(encoder_dev, buffer)) {
+		LOG_ERR("Unable to enqueue encoder output buf");
+		return -1;
+	}
```
How about contributing to samples transition to Zephyr-like coding style for new code, so that the original error code can be returned?

```c
	ret = (video_enqueue(encoder_dev, buffer);
	if (ret < 0) {
		LOG_ERR("Unable to enqueue encoder output buf");
		return ret;
	}
```

josuah reviewed (commented):
I've spotted one extra issue.
Waiting the comments from @jfischer-no to be resolved otherwise LGTM,
Thanks!

Commit:
samples: usb: uvc: add indirection for UVC source device

In preparation for the introduction of video encoder support
add an indirection for handling of the buffers of the UVC
source device. Currently this is only video_dev however it
can also be an encoder device when encoder is introduced
between video capture device and the UVC device.

Signed-off-by: Alain Volmat <alain.volmat@foss.st.com>

Commit:
usb: uvc: use struct uvc_frame_descriptor in most of function parameters

This commit prepares introduction of the UVC Frame Based support by
using the struct uvc_frame_descriptor as parameter of most of the UVC
functions. struct uvc_frame_descriptor contains the common fields for
all supported frame type and then depending on the DescriptorSubtype
the pointer is casted in the correct struct definition.

Signed-off-by: Alain Volmat <alain.volmat@foss.st.com>

Commit:
video: add H264 estimate in video_estimate_fmt_size

Add rough estimate of a worth case H264 output size.

The video_estimate_fmt_size would need more information
such as quality, profile in order to give a better
estimate for each formats so for the time being just
stick to 16bpp based size, same as for JPEG.

Signed-off-by: Alain Volmat <alain.volmat@foss.st.com>

avolmat-st commented:
I push the update of the PR. This also include the update done and tracked with the PR #93192 and PR #94562.

Only the last 10 commits of this PR are really new in this PR. One last modification (if that's the way we want to go) will be needed, which is to add the zephyr_h264enc / zephyr_jpegenc label in supported plateforms, which is currently only the STM32N6 I think. Following commit would be added, however this is not part of this PR since it isn't yet rebased on latest head which include those node:

```diff
diff --git a/boards/st/stm32n6570_dk/stm32n6570_dk_common.dtsi b/boards/st/stm32n6570_dk/stm32n6570_dk_common.dtsi
index 0f1096116a1..2541126f198 100644
--- a/boards/st/stm32n6570_dk/stm32n6570_dk_common.dtsi
+++ b/boards/st/stm32n6570_dk/stm32n6570_dk_common.dtsi
@@ -518,10 +518,10 @@ csi_interface: &dcmipp {
        };
 };
 
-&venc {
+zephyr_h264enc: &venc {
        status = "okay";
 };
 
-&jpeg {
+zephyr_jpegenc: &jpeg {
        status = "okay";
 };

```

etienne-lms reviewed (commented):
For commit "usb: device_next: uvc: allow UVC to enumerate when out of descriptors":
the commit message should also mention it fixes `uvc_add_vs_frame_desc()` that did not check `uvc_add_vs_frame_interval()` return value.

etienne-lms commented on drivers/video/video_stm32_dcmipp.c:
```diff
@@ -1359,22 +1359,95 @@ static int stm32_dcmipp_dequeue(const struct device *dev, struct video_buffer **
+static const struct video_format_cap stm32_dcmipp_dump_fmt[] = {
+	{
+		.pixelformat =
+			VIDEO_FOURCC_FROM_STR(CONFIG_VIDEO_STM32_DCMIPP_SENSOR_PIXEL_FORMAT),
```
Nitpicking again, fits in a single line:
```c
		.pixelformat = VIDEO_FOURCC_FROM_STR(CONFIG_VIDEO_STM32_DCMIPP_SENSOR_PIXEL_FORMAT),

etienne-lms commented on include/zephyr/drivers/video.h:
```diff
@@ -1003,6 +1003,24 @@ int video_estimate_fmt_size(struct video_format *fmt);
+/**
+ * @brief Transfer a buffer between 2 video device
+ *
+ * Helper function which dequeue a buffer from a source device and enqueue into a
```
```suggestion
 * Helper function which dequeues a buffer from a source device and enqueues into a

etienne-lms reviewed (commented):
Some minor comments.

josuah commented on subsys/usb/device_next/class/usbd_uvc.c:
```diff
@@ -1624,6 +1642,7 @@ static int uvc_init(struct usbd_class_data *const c_data)
 	struct video_caps caps;
 	uint32_t prev_pixfmt = 0;
 	uint32_t mask = 0;
+	int first_err = ret;
```
Good catch, I fixed it in the wrong commit.

josuah commented on subsys/usb/device_next/class/usbd_uvc.c:
```diff
@@ -1551,7 +1548,7 @@ static int uvc_add_vs_frame_desc(const struct device *dev,
 	desc->wWidth = sys_cpu_to_le16(w);
 	desc->wHeight = sys_cpu_to_le16(h);
-	desc->dwMaxVideoFrameBufferSize = sys_cpu_to_le32(max_size);
+	desc->dwMaxVideoFrameBufferSize = sys_cpu_to_le32(fmt.size);
```
Yes, [here](https://github.com/tinyvision-ai-inc/zephyr/blob/7ef196b9ee60ddd8458204c2c38d7eecd281f2a0/samples/subsys/usb/uvc/src/main.c#L62):

```c
	/* Set the format to get the size */
	ret = video_set_format(video_dev, &fmt);
	if (ret != 0) {
		LOG_ERR("Could not set the format of %s to %s %ux%u (size %u)",
			video_dev->name, VIDEO_FOURCC_TO_STR(fmt.pixelformat),
			fmt.width, fmt.height, fmt.size);
		return;
	}
```

It works like a handshake: setting the format to get the corresponding size back, as the driver expects.

I added a note about to the docstring as not obvious.

josuah replied:
Note: the PR for it is in https://github.com/zephyrproject-rtos/zephyr/pull/93192

avolmat-st commented on samples/drivers/video/tcpserversink/src/main.c:
```diff
@@ -422,7 +429,11 @@ int main(void)
+			ret = video_enqueue(video_dev, buffers[i]);
+			if (ret) {
+				LOG_ERR("Unable to enqueue video buf");
+				return 0;
```
Overall in this sample the error code are either return 0 or -1 or ret depending on the function. I think this would be fixed by the rework of video related sample app ongoing in PR #96072 so here I return the same kind of error code as return by other calls "around". That's not good, but overall correction as to be done, which is not in the scope of this PR.

Commit:
samples: usb: uvc: add h264/jpeg suffixed overlay

Add overlay files in order to enable usage of
the encoder in the UVC sample.
This work with platform defining node label
    zephyr_jpegenc
    zephyr_h264enc

Mode can be selected by using -DFILE_SUFFIX="jpegenc" or
-DFILE_SUFFIX="h264enc" when building the sample while
also adding -DCONFIG_VIDEO_ENCODER_JPEG or
-DCONFIG_VIDEO_ENCODER_H264 as well in the command line.

Signed-off-by: Alain Volmat <alain.volmat@foss.st.com>

avolmat-st commented on samples/subsys/usb/uvc/src/main.c:
```diff
@@ -138,19 +166,124 @@ static void app_add_filtered_formats(void)
 	}
 }
 
+static int app_init_videoenc(const struct device *const dev)
```
Fixed as describe in comment just above.

josuah commented on subsys/usb/device_next/class/usbd_uvc.c:
```diff
@@ -542,22 +543,41 @@ static int uvc_get_vs_probe_frame_interval(const struct device *dev, struct uvc_
+	if (frame_desc->bDescriptorSubtype == UVC_VS_FRAME_UNCOMPRESSED ||
+	    frame_desc->bDescriptorSubtype == UVC_VS_FRAME_MJPEG) {
+		struct uvc_frame_discrete_descriptor *desc =
+			(struct uvc_frame_discrete_descriptor *)frame_desc;
```
To try to save one review cycle, I think there was a proposal to use `(void *)`:
```suggestion
		struct uvc_frame_discrete_descriptor *desc = (void *)frame_desc;
```

avolmat-st replied:
Ok, fixing it.

josuah commented on subsys/usb/device_next/class/usbd_uvc.c:
```diff
@@ -542,22 +543,41 @@ static int uvc_get_vs_probe_frame_interval(const struct device *dev, struct uvc_
 	switch (request) {
 	case UVC_GET_MIN:
-		probe->dwFrameInterval = sys_cpu_to_le32(frame_desc->dwFrameInterval[0]);
+		probe->dwFrameInterval = min;
```
To illustrate:

```suggestion
		probe->dwFrameInterval = sys_cpu_to_le32(min);
```

josuah commented on subsys/usb/device_next/class/usbd_uvc.c:
```diff
@@ -620,6 +640,15 @@ static int uvc_get_vs_format_from_desc(const struct device *dev, struct video_fo
+
+		LOG_DBG("Found descriptor for format %u, frame %u, GUID '%.4s', pixfmt %04x",
+			format_frame_based_desc->bFormatIndex, frame_desc->bFrameIndex,
+			format_frame_based_desc->guidFormat, fmt->pixelformat);
```
Feel also free to log the guidFormat to all other log strings if you prefered.

josuah commented on subsys/usb/device_next/class/usbd_uvc.c:
```diff
@@ -1502,6 +1545,19 @@ static int uvc_add_vs_frame_interval(struct uvc_frame_descriptor *const desc,
+
+		if (frame_desc->bFrameIntervalType >= CONFIG_USBD_VIDEO_MAX_FRMIVAL) {
+			LOG_WRN("Out of frame interval fields");
+			return -ENOSPC;
```
:x: this one is blocking (i.e. avoids bugs)

avolmat-st replied:
This seems to be the only occurence left. Changed to -ENOMEM.

josuah reviewed (changes requested):
I think I caught 2 small issues, and found a few things that you might be curious about but not causing issues.

Thanks, getting close!

Commit:
boards: stm32n6570_dk: add zephyr_h264enc / zephyr_jpegenc labels

Add zephyr_h264enc and zephyr_jpegenc labels on node in order to
be able to use VENC and JPEG codec from samples.

Signed-off-by: Alain Volmat <alain.volmat@foss.st.com>

avolmat-st commented on subsys/usb/device_next/class/usbd_uvc.c:
```diff
@@ -1502,6 +1545,19 @@ static int uvc_add_vs_frame_interval(struct uvc_frame_descriptor *const desc,
+			(struct uvc_frame_based_discrete_descriptor *)desc;
+
+		if (frame_desc->bFrameIntervalType >= CONFIG_USBD_VIDEO_MAX_FRMIVAL) {
+			LOG_WRN("Out of frame interval fields");
```
Yes, kept same message as in case of UNCOMPRESSED/MJPEG

Commit:
samples: usb: uvc: add h264 / jpeg encoder test entries

Add entries in sample.yaml for enabling h264enc / jpegenc
uvc based test on the stm32n6570_dk/stm32n657xx/sb platform.

Signed-off-by: Alain Volmat <alain.volmat@foss.st.com>

josuah commented on drivers/video/video_common.c:
This looks like nothing but it saves a lot of variable definition in the samples :+1:

ngphibang commented on include/zephyr/drivers/video.h:
```diff
@@ -1003,6 +1003,24 @@ int video_estimate_fmt_size(struct video_format *fmt);
+ *
+ * @return 0 on success, otherwise a negative errno code
+ */
+int video_transfer_buffer(const struct device *src, const struct device *sink,
```
Need to be documented in release note ?

ngphibang commented on drivers/video/video_common.c:
```diff
@@ -456,6 +456,11 @@ int video_estimate_fmt_size(struct video_format *fmt)
+	case VIDEO_PIX_FMT_H264:
+		/* Rough estimate for the worst case */
+		fmt->pitch = 0;
+		fmt->size = fmt->width * fmt->height * 2;
```
Yes, it's way too much for H.264 (equivalent to raw uncompressed image frame). I think for H.264 it depends a number of parameters like profile, bitrate, etc. which cannot be passed to this function (where we have only width / height)

ngphibang replied:
So, you will have a better estimate or fall these two cases ?

avolmat-st replied:
I don't have now and without adding more parameters to this function that doesn't seem achievable so, let's put both cases together for now.

avolmat-st commented:
Had to push a new version to correct an issue in the ```app_add_filtered_formats``` function when doing iteration over the caps of the encoder & camera receiver. Code slightly reworked to avoid potential access to wrong index of caps table while still looking properly at the resolution exposed by the camera receiver. Thank you @josuah for finding the issue and working together to figure out an acceptable solution.

josuah reviewed (dismissed):
Thanks for finding the a simple solution that works for all current supported hardware!

jfischer-no commented on subsys/usb/device_next/class/usbd_uvc.c:
```diff
@@ -626,6 +638,15 @@ static int uvc_get_vs_format_from_desc(const struct device *dev, struct video_fo
 			format_desc->bFormatIndex, frame_desc->bFrameIndex);
+	} else if (format_desc->bDescriptorSubtype == UVC_VS_FORMAT_FRAME_BASED) {
+		struct uvc_format_frame_based_descriptor *format_frame_based_desc =
+			(void *)format_desc;
```
Like above just
		`struct uvc_frame_based_discrete_descriptor *desc = (void *)format_desc;`

avolmat-st commented on subsys/usb/device_next/class/usbd_uvc.c:
```diff
@@ -1475,21 +1486,28 @@ static void uvc_set_vs_bitrate_range(struct uvc_frame_discrete_descriptor *const
 {
-	int i = desc->bFrameIntervalType;
+	if (desc->bDescriptorSubtype == UVC_VS_FRAME_UNCOMPRESSED ||
+	    desc->bDescriptorSubtype == UVC_VS_FRAME_MJPEG) {
```
On the first patch, yes it would be simpler, avoid to add a level of indentation as well. However on the next commit when adding FRAME_BASED, the if becomes like below, hence an  if else or switch would still be needed.

```c
        if (desc->bDescriptorSubtype == UVC_VS_FRAME_UNCOMPRESSED ||
            desc->bDescriptorSubtype == UVC_VS_FRAME_MJPEG) {
                struct uvc_frame_discrete_descriptor *frame_desc = (void *)desc;

                if (frame_desc->bFrameIntervalType >= CONFIG_USBD_VIDEO_MAX_FRMIVAL) {
                        LOG_WRN("Out of descriptors, raise CONFIG_USBD_VIDEO_MAX_FRMIVAL above %u",
                                CONFIG_USBD_VIDEO_MAX_FRMIVAL);
                        return -ENOMEM;
                }

                frame_desc->dwFrameInterval[frame_desc->bFrameIntervalType] =
                        sys_cpu_to_le32(video_frmival_nsec(frmival) / 100);
                frame_desc->bFrameIntervalType++;
                frame_desc->bLength += sizeof(uint32_t);
        } else if (desc->bDescriptorSubtype == UVC_VS_FRAME_FRAME_BASED) {
                struct uvc_frame_based_discrete_descriptor *frame_desc = (void *)desc;

                if (frame_desc->bFrameIntervalType >= CONFIG_USBD_VIDEO_MAX_FRMIVAL) {
                        LOG_WRN("Out of descriptors, raise CONFIG_USBD_VIDEO_MAX_FRMIVAL above %u",
                                CONFIG_USBD_VIDEO_MAX_FRMIVAL);
                        return -ENOMEM;
                }

                frame_desc->dwFrameInterval[frame_desc->bFrameIntervalType] =
                        sys_cpu_to_le32(video_frmival_nsec(frmival) / 100);
                frame_desc->bFrameIntervalType++;
                frame_desc->bLength += sizeof(uint32_t);
        } else {
                LOG_DBG("Invalid frame type");
                return -EINVAL;
        }
```

jfischer-no commented on samples/subsys/usb/uvc/src/main.c:
```diff
@@ -38,7 +62,7 @@ static bool app_is_supported_format(uint32_t pixfmt)
 static bool app_has_supported_format(void)
 {
-	const struct video_format_cap *fmts = video_caps.format_caps;
+	const struct video_format_cap *fmts = app_uvc_source_caps()->format_caps;
```
I would appreciate it if you could make it a bit simpler and also use `*const`.
```c
const struct video_caps *const caps = app_uvc_source_caps();
const struct video_format_cap *const fmts = caps->format_caps;
```
